### PR TITLE
Add nullable type to theory of datatypes

### PR DESCRIPTION
**Pull request by mudathirmahgoub**

### CODE REVIEW

File: src/expr/node_manager_template.cpp [hunk 1]
```diff
@@ -941,6 +958,39 @@ TypeNode NodeManager::mkTupleType(const std::vector<TypeNode>& types)
...
+    return it->second;
+  }
+  // construct the corresponding datatype with two constructors
+  // __cvc5_nullable_type_ctor_null, and __cvc5_nullable_type_ctor_some
+  std::stringstream sst;
+  sst << "__cvc5_nullable_" << type;
+  DType dt(sst.str());
+  dt.setNullable();
+  std::stringstream sscNull;
+  sscNull << sst.str() << "_ctor_null";
```
Comment (ajreynol) [inline src/expr/node_manager_template.cpp:976]: You should just be able to give this its proper name, i.e. `nullable.null`.  This should also avoid the need for some of the special casing in the smt2 printer.

File: src/expr/node_manager_template.cpp [hunk 2]
```diff
@@ -941,6 +958,39 @@ TypeNode NodeManager::mkTupleType(const std::vector<TypeNode>& types)
...
+  std::shared_ptr<DTypeConstructor> null =
+      std::make_shared<DTypeConstructor>(sscNull.str());
+  dt.addConstructor(null);
+  std::shared_ptr<DTypeConstructor> some =
+      std::make_shared<DTypeConstructor>(sscNull.str());
+  std::stringstream sscSome;
+  sscSome << sst.str() << "_ctor_some";
```
Comment (ajreynol) [inline src/expr/node_manager_template.cpp:983]: Ditto, call this `nullable.some`.

File: src/parser/smt2/smt2_state.cpp [hunk 3]
```diff
@@ -159,6 +159,15 @@ void Smt2State::addDatatypesOperators()
     defineVar("tuple.unit", d_solver->mkTuple({}));
     addIndexedOperator(Kind::UNDEFINED_KIND, "tuple.select");
     addIndexedOperator(Kind::UNDEFINED_KIND, "tuple.update");
+    Sort btype = d_solver->getBooleanSort();
+    defineVar("nullable.null",
+              d_solver->mkNullableNull(d_solver->mkNullableSort(btype)));
+    addOperator(Kind::APPLY_CONSTRUCTOR, "nullable.some");
+    addOperator(Kind::APPLY_CONSTRUCTOR, "nullable.val");
+    addOperator(Kind::NULLABLE_LIFT, "nullable.lift");
+    addOperator(Kind::APPLY_CONSTRUCTOR, "nullable.is_null");
```
Comment (ajreynol) [inline src/parser/smt2/smt2_state.cpp:168]: Can you make this `APPLY_TESTER`? Ditto for `is_some`.

File: src/parser/smt2/smt2_state.cpp [hunk 4]
```diff
@@ -284,6 +293,7 @@ void Smt2State::addCoreSymbols()
 {
   defineType("Bool", d_solver->getBooleanSort(), true);
   Sort tupleSort = d_solver->mkTupleSort({});
+  defineType("Nullable", d_solver->mkNullableSort(tupleSort), true);
```
Comment (ajreynol) [inline src/parser/smt2/smt2_state.cpp:296]: This shouldn't be necessary. Can you try removing this line?

File: src/expr/node_manager_template.cpp [hunk 5]
```diff
@@ -941,6 +958,39 @@ TypeNode NodeManager::mkTupleType(const std::vector<TypeNode>& types)
...
+  some->addArg("_stor", type);
```
Comment (ajreynol) [inline src/expr/node_manager_template.cpp:984]: Ditto call this `nullable.val`.

File: test/regress/cli/regress1/datatypes/nullables/nullables2.smt2 [hunk 6]
```diff
@@ -0,0 +1,8 @@
+(set-logic ALL)
+(set-info :status unsat)
```
Comment (ajreynol) [inline test/regress/cli/regress1/datatypes/nullables/nullables2.smt2:2]: All `unsat` benchmarks with `Nullable` will fail until we add them to the proof signatures for ALF/LFSC.

For now, add `; DISABLE-TESTER: proof` to this benchmark (and any other unsat ones with Nullable).

**Review by ajreynol — CHANGES_REQUESTED** [review]

File: test/regress/cli/regress1/datatypes/nullables/nullables2.smt2 (hunk 6 above)
Comment (mudathirmahgoub) [reply test/regress/cli/regress1/datatypes/nullables/nullables2.smt2:3]: Done

File: src/parser/smt2/smt2_state.cpp (hunk 4 above)
Comment (mudathirmahgoub) [reply src/parser/smt2/smt2_state.cpp]: Done.

File: src/expr/node_manager_template.cpp (hunk 5 above)
Comment (mudathirmahgoub) [reply src/expr/node_manager_template.cpp]: Done.

File: src/expr/node_manager_template.cpp (hunk 2 above)
Comment (mudathirmahgoub) [reply src/expr/node_manager_template.cpp]: Done.

File: src/expr/node_manager_template.cpp (hunk 1 above)
Comment (mudathirmahgoub) [reply src/expr/node_manager_template.cpp]: Done.

File: src/parser/smt2/smt2_state.cpp (hunk 3 above)
Comment (mudathirmahgoub) [reply src/parser/smt2/smt2_state.cpp]: Done

File: src/printer/smt2/smt2_printer.cpp [hunk 7]
```diff
@@ -846,20 +864,40 @@ bool Smt2Printer::toStreamBase(std::ostream& out,
...
       Node op = n.getOperator();
       size_t cindex = DType::indexOf(op);
       const DType& dt = DType::datatypeOf(op);
-      out << "(_ is ";
-      toStream(out,
-               dt[cindex].getConstructor(),
-               lbind,
-               toDepth < 0 ? toDepth : toDepth - 1);
-      out << ")";
+      if (n[0].getType().isNullable())
```
Comment (ajreynol) [inline src/printer/smt2/smt2_printer.cpp:879]:
```suggestion
      if (dt.isNullable())
```

**Review by ajreynol — CHANGES_REQUESTED** [review]
Looks good.

You'll need to add unit tests to cover all the new API methods.

**Comment by ajreynol** [comment]
Also, can you add to NEWS.md? We also should update the website documentation (in a later PR).

**Comment by mudathirmahgoub** [comment]
> Looks good.
> 
> You'll need to add unit tests to cover all the new API methods.

Done

**Comment by mudathirmahgoub** [comment]
> Also, can you add to NEWS.md? We also should update the website documentation (in a later PR).

Done.

**Comment by ajreynol** [comment]
You need to add the Python API and tests as well.

**Comment by mudathirmahgoub** [comment]
> You need to add the Python API and tests as well.

Done.

File: src/printer/smt2/smt2_printer.cpp [hunk 8]
```diff
@@ -529,8 +530,13 @@ bool Smt2Printer::toStreamBase(std::ostream& out,
           out << " ";
           toStreamType(out, dt[0][i].getRangeType());
         }
-        out << ")";
+        out << ")";
```
Comment (ajreynol) [inline src/printer/smt2/smt2_printer.cpp:533]: Will need to clang-format.

**Review by ajreynol — APPROVED** [review]
Be sure to clang-format.

File: src/printer/smt2/smt2_printer.cpp (hunk 8 above)
Comment (mudathirmahgoub) [reply src/printer/smt2/smt2_printer.cpp]: Fixed

**Pull request merged**